### PR TITLE
v1.5.0 - Performance improvements & rewrite

### PR DESCRIPTION
**Pull request by GoatG33k**
- Bumped version to 1.5.0
- Separated client code into separate files
- Added a new configuration option for the update rate
- 50% performance improvement (~0.08ms/frame -> ~0.04ms/frame)
- Cleaned up and simplified code

### CODE REVIEW

File: cl_render.lua [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    local formatTemplate = config.text.format
+    local _total = #postals
+
+    while true do
+        local coords = GetEntityCoords(PlayerPedId())
+        local _nearestIndex, _nearestD
+        coords = vec(coords[1], coords[2])
+
+        for i = 1, _total do
+            local D = #(coords - postals[i][1])
```
Comment (blockba5her) [inline cl_render.lua:38]: This internally performs a `sqrt`, so it's probably better to switch over to checking distance magnitudes in these loops. You can either use the built in function `Vdist2` or writing something like:
```lua
local v = coords - postals[i][1]
local D = v.x^2 + v.y^2 + v.z^2
```

then after the loop you can find the `sqrt` for finding canonical distance from the postal

File: config.lua [hunk 2]
```diff
@@ -1,21 +1,45 @@
...
+        color = 3, -- default 3 (light blue)
+        distToDelete = 100.0, -- in meters
+        deleteText = 'Route deleted',
+        drawRouteText = 'Drawing a route to %s',
+        notExistText = "That postal doesn't exist"
+    },
+
+    --- The delay in milliseconds until the current postal code and text are updated
+    --- I wouldn't recommend anything lower than 50ms, as calculating the postal can be intensive.
+    updateDelay = 300,
```
Comment (blockba5her) [inline config.lua:44]: It's probably for the best to remove the classification comments from the configuration file as I'm sure people would probably get confused. First hand experience dealing with them lol

File: fxmanifest.lua [hunk 3]
```diff
@@ -8,19 +8,27 @@ If you just want to change the postal file, **ONLY** change the above variable
...
-	'config.lua',
-	'cl.lua'
+    'config.lua',
+
+    'cl.lua',
+    'cl_commands.lua',
+    'cl_render.lua',
+
+    -- uncomment to enable dev tools
+    --'cl_dev.lua',
```
Comment (blockba5her) [inline fxmanifest.lua:26]: This is fine, but please update the `README.md` with the new instructions for enabling dev mode.

**Review by blockba5her — CHANGES_REQUESTED** [review]

File: cl.lua [hunk 4]
```diff
@@ -1,200 +1,19 @@
...
-	)
-end
-
-exports('getPostal', function() if nearest ~= nil then return postals[nearest.i].code else return nil end end)
+---@class PostalData : table<number, vec>
+---@field code string
+---@type table<number, PostalData>
+postals = LoadResourceFile(GetCurrentResourceName(), GetResourceMetadata(GetCurrentResourceName(), 'postal_file'))
+postals = json.decode(postals)
+for i, postal in ipairs(postals) do postals[i] = { vec(postal.x, postal.y), code = postal.code } end
```
[the remark this answers is not in the document]
Comment (GoatG33k) [reply cl.lua:6]: Can do 👍, I'll make all of these changes tonight.

File: cl_render.lua [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+local IsHudHidden = IsHudHidden
+local SetTextFont = SetTextFont
+local CreateThread = CreateThread
+local SetTextScale = SetTextScale
+local SetTextOutline = SetTextOutline
+local GetEntityCoords = GetEntityCoords
+local EndTextCommandDisplayText = EndTextCommandDisplayText
+local BeginTextCommandDisplayText = BeginTextCommandDisplayText
+local AddTextComponentSubstringPlayerName = AddTextComponentSubstringPlayerName
+-- end optimizations
```
[the remark this answers is not in the document]
Comment (GoatG33k) [reply cl_render.lua:16]: Localized variables are faster than global accesses in Lua ([as described in this document](https://www.lua.org/gems/sample.pdf)). In this case, I only localized natives in this function as they are used repeatedly in render loops, but the performance improvement is measurable.

File: config.lua (hunk 2 above)
Comment (GoatG33k) [reply config.lua:44]: I was pondering this as well, I think I will remove the luadoc in favor of individual comments for each. I will also handle fallback support for users with old `config.lua` files. Does `300ms` seem like an appropriate default to you though?

File: cl_render.lua (hunk 1 above)
Comment (GoatG33k) [reply cl_render.lua:38]: I'm not sure I understand what you mean by this, could you clarify why `sqrt` is necessary at all?

The `#(vecA - vecB)` syntax does an optimized distance calculation, and should return the distance in game units (meters). `Vdist2` and custom math will likely be slower than native Lua implemented in C.

**Comment by GoatG33k** [comment]
I'd also like to thank you for actually reviewing this massive PR @blockba5her 💪

File: cl_render.lua (hunk 5 above)
[the remark this answers is not in the document]
Comment (blockba5her) [reply cl_render.lua:16]: Never knew, but makes sense when explained.

File: cl_render.lua (hunk 1 above)
Comment (blockba5her) [reply cl_render.lua:38]: The point that I'm trying to make is that a `sqrt` wouldn't be necessary for comparisons between distances, only for the final display of distance. It's explained pretty well on the [C# Vector3 Docs](https://github.com/citizenfx/fivem/blob/master/code/client/clrcore/Math/Vector3.cs#L649) why you would want to use distance magnitude over distance.

Do some testing, and see if using `#(vec1 - vec2)` is faster than writing it out in lua or using `Vdist2`. Or maybe it doesn't make any difference at all :shrug:

File: config.lua (hunk 2 above)
Comment (blockba5her) [reply config.lua:44]: `300ms` works for me 👍

**Comment by blockba5her** [comment]
> I'd also like to thank you for actually reviewing this massive PR @blockba5her 💪

I was discouraged at first, but I saw that you somehow doubled the performance and I just had to know how lol

File: cl_render.lua (hunk 1 above)
Comment (GoatG33k) [reply cl_render.lua:38]: I'll do some more in-depth performance testing and get back to you!

**Comment by GoatG33k** [comment]
I apologize, I've been a bit busy lately. I'm just getting around to the performance testing now -- I still would like to get this merged 😄

File: cl_render.lua (hunk 1 above)
Comment (GoatG33k) [reply cl_render.lua:38]: Here are the results I got from my performance testing. These times do include the network time (as I used events), but it should be negligible for localhost. (The repo of my perf testing code is [available here](https://github.com/GoatG33k/FiveM-dist-calc-perf-testing))

| **Code**                                                 | **Iterations** | **Mean Time** | **Mean ops/sec**  |
|----------------------------------------------------------|----------------|---------------|-------------------|
| `local D = #(vecA - vecB)`                               | 5x @ 10mil     | 447.6ms       | ~22,388.4 ops/sec |
| `local v = vecA - vecB; local D = v.x^2 + v.y^2 + v.z^2` | 5x @ 10mil     | 726.0ms       | ~13,783.8 ops/sec |
| `local D = Vdist(a[1], a[2], a[3], b[1], b[2], b[3])`    | 5x @ 10mil     | 1682.0ms      | ~5,950.4 ops/sec  |
| `local D = Vdist2(a[1], a[2], a[3], b[1], b[2], b[3])`   | 5x @ 10mil     | 1635.0ms      | ~6,139.4 ops/sec  |

File: cl.lua (hunk 4 above)
[the remark this answers is not in the document]
Comment (GoatG33k) [reply cl.lua]: I couldn't find anything about `CreateThreadNow`, but I did make it run in a thread, and the render thread waits for it to become available. 😄

**Review by blockba5her — APPROVED** [review]
Changes LGTM.

**Pull request merged**